types/wlr_surface_layers.c

```diff
@@ -0,0 +1,212 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/types/wlr_surface.h>
+#include <wlr/types/wlr_surface_layers.h>
+#include <wlr/util/log.h>
+#include "surface-layers-protocol.h"
+
+static const char *surface_layers_role = "layer_surface";
+
+static void layer_surface_set_interactivity(struct wl_client *client,
+		struct wl_resource *resource, uint32_t input_types,
+		uint32_t exclusive_types) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->input_types = input_types;
+	surface->exclusive_types = exclusive_types;
+	wl_signal_emit(&surface->events.set_interactivity, surface);
+}
+
+static void layer_surface_set_anchor(struct wl_client *client,
+		struct wl_resource *resource, uint32_t anchor) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->anchor = anchor;
```

Comment:
Maybe a protocol error for invalid anchors (like orthogonal anchors or anchors outside the enum).

Reply:
Again, I don't know: the rest of the wayland protocol isn't sending errors for things like this (e.g. [`edges` in `wl_shell_surface::resize`](https://wayland.freedesktop.org/docs/html/apa.html#protocol-spec-wl_shell_surface-request-resize)).

Reply:
Many of them do this, like xdg-shell and data-device.

Reply:
Protocol error here. And we should put them everywhere.